src/Microsoft.Diagnostics.Monitoring.EventPipe/Logs/EventLogsPipeline.cs

```diff
@@ -110,14 +115,29 @@ protected override Task OnEventSourceAvailable(EventPipeEventSource eventSource,
                 ILogger logger = _factory.CreateLogger(categoryName);
                 List<IDisposable> scopes = new();
 
-                if (logActivities.TryGetValue(traceEvent.ActivityID, out LogActivityItem logActivityItem))
+                //
+                // The MessageJson event will occur with an ActivityId equal to the most relevant activity branch and we can backtrack to the root of the tree
```

Comment:
Here's a visual representation of the problem.
- Node `D` is caused by another trace with an event start. Our code is unaware of it and so when scope `E` occurs, it'll effectively be the root node of its own tree.
- If a log message occurs on node `D` then we can use the `RelatedActivityId` to find our way back to node `B`. NOTE: This will only work if we're the first event message to happen in this branch.


```mermaid
graph TD
    A[A: scope] ---|RelatedActivityId| B[B: scope]
    B --- C[C: scope]
    B -.-D[D: unknown]
    D -.- E[E: scope]
    E --- F[F: scope]
```

Reply:
There's a few potential solutions to this (but these should be considered as a separate effort):
1. Enhance Tpl to not use a singleton for it's activity tracker, making it session-aware.
2. Enhance the EventSourceLogger with additional information that we can use to construct this graph without relying on Tpl.
3. Hook Tpl's `DebugFacilityMessage` and `DebugFacilityMessage1` events. These will broadcast when an ActivityId branches (or the branch gets pruned). However I have concerns around performance with this as these are general-purpose debugging events not specific to branching alone. Some performance analysis would need to be done.
4. Implemented a multiplexing trace session for dotnet-monitor instead of using multiple individual sessions. This would let us be aware of other event starts that occur.